codex/slots/proofs/backends/circomcompat.nim

```diff
@@ -143,58 +182,139 @@ proc prove[H](self: CircomCompat, input: NormalizedProofInputs[H]): ?!CircomProo
+      else:
+        return failure(exc.msg)
+
+  if not task.success.load():
+    return failure("Failed to prove circuit")
+
+  success()
+
+proc prove*[H](
+    self: CircomCompat, input: ProofInputs[H]
+): Future[?!CircomProof] {.async, raises: [CancelledError].} =
+  var proof = newProof()
+  defer:
+    destroyProof(proof)
+
+  try:
+    if error =? (await self.asyncProve(self.normalizeInput(input), proof)).errorOption:
+      return failure(error)
+    return success(deepCopy(proof)[])
+  except CancelledError as exc:
+    raise exc
+
+proc circomVerifyTask(task: ptr VerifyTask) {.gcsafe.} =
+  defer:
+    task[].inputs[].releaseCircomInputs()
+    discard task[].signal.fireSync()
+  let res = verify_circuit(task[].proof, task[].inputs, task[].vkp)
+  if res == ERR_OK:
+    task[].success[].store(true)
+  elif res == ERR_FAILED_TO_VERIFY_PROOF:
+    task[].success[].store(false)
+  else:
+    task[].success[].store(false)
+    error "Failed to verify proof", errorCode = res
+
+proc asyncVerify*[H](
+    self: CircomCompat,
+    proof: CircomProof,
+    inputs: ProofInputs[H],
+    success: ptr Atomic[bool],
+): Future[?!void] {.async.} =
+  var proofPtr = unsafeAddr proof
+  var inputs = inputs.toCircomInputs()
+
+  without threadPtr =? ThreadSignalPtr.new():
+    return failure("Unable to create thread signal")
+
+  defer:
+    threadPtr.close().expect("closing once works")
+
+  var task = VerifyTask(
+    proof: proofPtr,
+    vkp: self.vkp,
+    inputs: addr inputs,
+    success: success,
+    signal: threadPtr,
+  )
+
+  let taskPtr = addr task
+
+  doAssert self.taskpool.numThreads > 1,
+    "Must have at least one separate thread or signal will never be fired"
 
-  success proof
+  self.taskpool.spawn circomVerifyTask(taskPtr)
 
-proc prove*[H](self: CircomCompat, input: ProofInputs[H]): ?!CircomProof =
-  self.prove(self.normalizeInput(input))
+  let threadFut = threadPtr.wait()
+
+  try:
+    await threadFut.join()
+  except CatchableError as exc:
+    try:
+      await threadFut
+    except AsyncError as asyncExc:
+      return failure(asyncExc.msg)
+    finally:
+      if exc of CancelledError:
+        raise (ref CancelledError) exc
+      else:
+        return failure(exc.msg)
+
+  success()
 
 proc verify*[H](
     self: CircomCompat, proof: CircomProof, inputs: ProofInputs[H]
-): ?!bool =
+): Future[?!bool] {.async, raises: [CancelledError].} =
   ## Verify a proof using a ctx
   ##
-
-  var
-    proofPtr = unsafeAddr proof
-    inputs = inputs.toCircomInputs()
-
+  var res = newVerifyResult()
+  defer:
+    destroyVerifyResult(res)
   try:
-    let res = verify_circuit(proofPtr, inputs.addr, self.vkp)
-    if res == ERR_OK:
-      success true
-    elif res == ERR_FAILED_TO_VERIFY_PROOF:
-      success false
-    else:
-      failure("Failed to verify proof - err code: " & $res)
-  finally:
-    inputs.releaseCircomInputs()
+    if error =? (await self.asyncVerify(proof, inputs, res)).errorOption:
```

Comment:
why not just return a bool here (or even void) and set to false on failure, true otherwise?

Reply:
The response is either passed or failed, right?